lib/asserts.nix

```
@@ -2,78 +2,122 @@
 
 rec {
 
-  /* Throw if pred is false, else return pred.
-     Intended to be used to augment asserts with helpful error messages.
+  /**
+    Throw if pred is false, else return pred.
+    Intended to be used to augment asserts with helpful error messages.
 
-     Example:
-       assertMsg false "nope"
-       stderr> error: nope
+    # Example
 
-       assert assertMsg ("foo" == "bar") "foo is not bar, silly"; ""
-       stderr> error: foo is not bar, silly
+    ```nix
+    assertMsg false "nope"
+    stderr> error: nope
+    assert assertMsg ("foo" == "bar") "foo is not bar, silly"; ""
+    stderr> error: foo is not bar, silly
+    ```
```

It's possible the examples aren't being counted anymore because this is not using the syntax we have for examples. Can you try the following syntax and see if it works?
(applies to other functions too)

```suggestion
    :::{.example}
    # `lib.assertMsg` usage examples
    ```nix
    assertMsg false "nope"
    stderr> error: nope
    assert assertMsg ("foo" == "bar") "foo is not bar, silly"; ""
    stderr> error: foo is not bar, silly
    ```
    :::
```

---

I'm not sure if nixos-render-docs will complain about the example not having a header attached to it (it's usually something like `:::{.example #ex-header-for-example-here}`).

If it complains, let's see how much work it is to make it auto-generate these for `lib` docs.

---

No the rendering doesn't complain. Its just that the class is not part of the comment.
And thinking of it, i think we shouldn't add those classes and let the rendering do it.

---

> i think we shouldn't add those classes and let the rendering automatically assign those

See [this example](https://nixos.org/manual/nixpkgs/unstable/#ex-dockerTools-streamLayeredImage-exploringlayers) for a more complicated case - the example itself contains paragraphs of explanation along with multiple code boxes.

If we want the rendering to automatically convert the content of the examples header into the `:::{.example ...}` syntax, I think we'll end up having to force some structure on the content of the examples header. My first idea is something like:

- Each example inside the examples header needs to have its own sub-header:
    ```
    # Examples
    ## `lib.assertMsg` usage examples
    ... content
    ## Some other example header here
    ... content
    ```
- The content of each sub-header will be included within a `:::{.example}` block (and the sub-header becomes the header for the example block)

However, IMO if we're forcing this kind of structure in the examples header to only then transform into `:::{.example}` blocks, might as well cut this step and just ask people to organise the examples inside `:::{.example}` blocks directly.

---

I suggest not adding another layer of magic and instead leaving it at the explicit `{.example}` blocks. The rationale is that otherwise you'll have to know yet another rule you must follow.